File: VRC.Car.Main/Messaging/MessagingHandler.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Net.Http;
using System.Net.Security;
using System.Net.Sockets;
using System.Threading.Tasks;
using Microsoft.AspNetCore.SignalR.Client;
using VRC.Shared.Car;

namespace VRC.Car.Main.Messaging
{
    public class MessagingHandler
    {
        private readonly string _hubUrl = "https://localhost:5001/racinghub";
        private HubConnection _hubConnection;

        public event EventHandler<CarCommandEventArgs> CarCommandReceivedEvent;

        public int CarNumber { get; set; }

        public MessagingHandler()
        {
            Initialise();
        }

        public bool IsConnected =>
        _hubConnection.State == HubConnectionState.Connected;

        private void Initialise()
        {
            // Create a handler that accepts custom (untrusted) certificates
            var handler = new HttpClientHandler
            {
                ClientCertificateOptions = ClientCertificateOption.Manual,
                ServerCertificateCustomValidationCallback = (httpRequestMessage, cert, cetChain, policyErrors) => {
                    // Validate the cert here and return true if it's correct.
                    // If this is a development app, you could just return true always
                    // In production you should ALWAYS either use a trusted cert or check the thumbprint of the cert matches one you expect.
                    return true;
                }
            };
            _hubConnection = new HubConnectionBuilder()
            .WithUrl(_hubUrl, options =>
            {
                // Register the custom handler above and also configure WebSockets
                options.HttpMessageHandlerFactory = _ => handler;
                options.WebSocketConfiguration = sockets =>
                {
                    sockets.RemoteCertificateValidationCallback = new RemoteCertificateValidationCallback((sender, certificate, chain, policyErrors) => {
                        // You have to repeat your cert validation code here. Feel free to use a helper method!
                        return true;
                    });
                };
            })
            .Build();

            _hubConnection.On<CarCommand>("ReceiveCarCommand", (command) =>
            {
                // Console.WriteLine($"Car number: {command.CarNumber} Car throttle: {command.Throttle} Car direction: {command.Direction}");
                CarCommandReceivedEvent?.Invoke(this, new CarCommandEventArgs(command));
            });

            _hubConnection.On<int>("AssignCarNumber", (carNumber) =>
            {
                CarNumber = carNumber;
                Console.WriteLine($"Our new car number: {carNumber}");
            });

            _hubConnection.Closed += async (error) =>
            {
                Console.WriteLine("Connection lost");
                await ConnectAsync();
            };

            _hubConnection.Reconnected += async (newConnectionId) =>
            {
                await _hubConnection.SendAsync("ReclaimCarNumber", CarNumber);
            };
        }

        public async Task ConnectAsync()
        {
            Console.WriteLine($"Connecting to API at \"{_hubUrl}\"...");
            while (!IsConnected)
            {
                try
                {
                    await _hubConnection.StartAsync();
                    Console.WriteLine("Connected to API");
                    await RequestCarNumberAsync();
                    while (CarNumber == 0)
                    {
                        await Task.Delay(50);
                    }
                }
                catch (System.Exception e)
                {
                    Console.WriteLine(e.Message);
                    Console.WriteLine(e.InnerException?.Message);
                    Console.WriteLine("Failed to connect to API");
                    Console.WriteLine("Retrying...");
                    await Task.Delay(2000);
                }
            }
        }

        // obsolete, for testing
        public async Task SendCarCommandAsync(int carNumber, CarCommand command) =>
            await _hubConnection.SendAsync("SendCarCommand", carNumber, command);

        private async Task RequestCarNumberAsync() =>
            await _hubConnection.SendAsync("RequestCarNumber");

        private string GetLocalIPAddress()
        {
            var host = Dns.GetHostEntry(Dns.GetHostName());
            foreach (var ip in host.AddressList)
            {
                if (ip.AddressFamily == AddressFamily.InterNetwork)
                {
                    return ip.ToString();
                }
            }
            throw new Exception("No network adapters with an IPv4 address in the system!");
        }
    }
}

File: VRC.Car.Pi/messaging_handler.py
from time import sleep
from signalrcore.hub_connection_builder import HubConnectionBuilder
from hardware_controller import HardwareController
import netifaces

class MessagingHandler:
  _hubUrl = 'https://192.168.0.201:5001/racinghub'
  _hubConnection = HubConnectionBuilder()\
    .with_url(_hubUrl, options={"verify_ssl": False}) \
    .build()
  _reconnect = False

  carNumber = 1

  def __init__(self, hardwareController):
    self._hardwareController = hardwareController
    self._hubConnection.on_open(self.on_connect)
    self._hubConnection.on_close(self.on_close)
    self._hubConnection.on_error(self.on_disconnect)
    self._hubConnection.on("ReceiveCarCommand", self.on_receive_car_command)
    self._hubConnection.on("AssignCarNumber", self.on_receive_car_number)

  def connect(self):
    print('Connecting to API at \"{}\"...'.format(self._hubUrl))
    self._hardwareController.display_text('Connecting to API')
    sleep(0.3)
    try:
      self._hubConnection.start()
    except:
      print('Failed to connect to API')
      self._hardwareController.display_text('Failed to connect to API')
      self._hardwareController.connection_status = 'error'
      sleep(0.3)

  def on_connect(self):
    print('Connected to API with ip \"{}\"'.format(self.get_ip_address()))
    self._hardwareController.connection_status = 'connected'
    if not self._reconnect:
      self.request_car_number()
    else:
      self.reclaim_car_number()

  def on_close(self):
    print("connection closed")
    self.stop_car()
    self._hardwareController.connection_status = 'disconnected'

  def on_disconnect(self):
    print("Connection lost")
    self.stop_car()
    self._hardwareController.connection_status = 'disconnected'
    self._reconnect = True
    self.connect()

  def on_receive_car_command(self, carCommand):
    self._hardwareController.send_car_command(carCommand[0])

  def stop_car(self):
    print('Stopping car')
    stopCommand = {
      'carNumber': self.carNumber,
      'direction': 'N',
      'throttle': 'N'
    }
    self._hardwareController.send_car_command(stopCommand)

  def on_receive_car_number(self, carNumber):
    print('Received car number')
    print('Our new car number: {}'.format(carNumber[0]))
    self.carNumber = carNumber[0]
    self._hardwareController.car_number = carNumber[0]

  def request_car_number(self):
    print('Requesting car number...')
    self._hubConnection.send("RequestCarNumber", [self.get_ip_address()])

  def reclaim_car_number(self):
    self._hubConnection.send("ReclaimCarNumber", [self.carNumber, self.get_ip_address()])

  def send_car_command(self):
    print('Sending car command to myself for testing...')
    command = {
      'carNumber': 1,
      'Direction': 'L',
      'Throttle': 'F'
    }
    self._hubConnection.send("SendCarCommand", [self.carNumber, command])

  def get_ip_address(self):
    ipAddress = netifaces.ifaddresses('wlan0')[2][0]['addr']
    print(ipAddress)
    return ipAddress
